---
title: "WM1119 vs DMSO (83, 96h)"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```


```{r}
setwd("~/RNA_seq_KAT6Ainhibitor_small_dataset/WM1119 vs DMSO (83, 96H)")

#Libraries and load in Data

library(tidyverse)
library(readxl)
library(dplyr)
library(pkgbuild) ##load pkgbuild
library(DESeq2) ##Load in Deseq library

countdata <-read_excel("KAT6A-inhibitors-controlNPC-CountMatrix_AN008D_221228 (1).xlsx")
  
countdata <- as.data.frame(countdata)
row.names(countdata) <- countdata[,1]
countdata <- countdata[,-1]

coldata <- data.frame(read_excel("KAT6A-inhibitors-controlNPC-colData-sampleINFO_AN008D_221228.xlsx"))

rownames(coldata)<-coldata[,1]
coldata<-coldata[,-1]

countdata<-data.frame(countdata)
```
```{r}
# Filter coldata for WM1119, DMSO, and the 888 cell line
filtered_coldata <- coldata %>%
  filter((Drug == "WM1119" | str_starts(Drug, "DMSO")) & Line == "83" & timept == "96H")

# Extract sample names from the filtered coldata
sample_names <- filtered_coldata$sampleID


# Select only the countdata columns that match these filtered sample names
countdata <- countdata %>%
  dplyr::select(all_of(sample_names))

coldata <- filtered_coldata
countdata <-countdata

view(countdata)
view(coldata)
```


```{r}

dds<-DESeqDataSetFromMatrix(countData = countdata, colData = coldata, design = ~ Drug)

dds<-DESeq(dds)

dispersion_graph <-plotDispEsts(dds)

```
```{r}
#make pca
rld<-rlog(dds,blind = TRUE)
#making fancy pca with labels 
#this is to plot the points on the PCA from DESeq2 with sample names on them
pcaData<-plotPCA(rld,intgroup=c("Drug", "Dosage"),returnData=TRUE)
percentVar<-round(100*attr(pcaData,"percentVar"))
pcaData<-merge(x=pcaData,y=coldata,by=0)
library(ggplot2)
library(ggrepel)
ggplot(pcaData, aes(PC1, PC2, color=Drug.x, shape=Dosage.x ,label=name)) +
  geom_point(size=3) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"))+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

```{r}
#get more than PC1 and PC2
#DESeq2 defaults to using the top 500 most variable gene 
morePCs<-function (object, intgroup =c("Drug", "Dosage"), ntop = 500, returnData = FALSE,firstPC,secondPC) 
{
  #object is the log2 transformed version of the gene count table 
  rv <- rowVars(assay(object))
  #pick out the ntop (default 500) genes with the most variable expression
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
                                                     length(rv)))]
  #subset the log2 transformed values to the ntop most variable gene expression and compute PCs
  pca <- prcomp(t(assay(object)[select, ]))
  #this variable is the percent explained by the chosen PCs
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, 
                                               drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = ":"))
  }
  else {
    colData(object)[[intgroup]]
  }
  d <- data.frame(PC1 = pca$x[, firstPC], PC2 = pca$x[, secondPC], group = group, 
                  intgroup.df, name = colnames(object))
  if (returnData) {
    attr(d, "percentVar") <- percentVar[firstPC:secondPC]
    return(d)
  }
  ggplot(data = d, aes_string(x = "PC1", y = "PC2", 
                              color = "group")) + geom_point(size = 3) + xlab(paste0("PC",firstPC,": ", 
                                                                                     round(percentVar[1] * 100), "% variance")) + ylab(paste0("PC",secondPC,": ", 
                                                                                                                                              round(percentVar[2] * 100), "% variance")) + coord_fixed()
}



```

```{r}
library(apeglm)

resultsNames(dds)
res<-lfcShrink(dds, coef= "Drug_WM1119_vs_DMSO",type="apeglm")

```

```{r}
gencode_genes<-read.csv("~/RNA_seq_KAT6Ainhibitor_small_dataset/gencode.v31.primary_assembly.genes.csv")

#gencode_genes<-read.csv("/Users/angelawei/Box Sync/_Arboleda Lab Stuff/personal/Angela_Wei/Reference_genome/GencodeV31_GR38/gencode.v31.primary_assembly.genes.csv")

gencode_genes<-dplyr::select(gencode_genes,gene_name,gene_type,seqnames,hgnc_id,gene_id)

#view(gencode_genes)
hcng_ref<-read.delim("~/RNA_seq_KAT6Ainhibitor_small_dataset/HCNG_05-09-19.txt")

#hcng_ref<-read.delim("/Users/angelawei/Box Sync/_Arboleda Lab Stuff/personal/Angela_Wei/Reference_genome/HCNG/HCNG_05-09-19.txt")
hcng_ref<-dplyr::select(hcng_ref,HGNC.ID,Approved.name,Chromosome,RefSeq.IDs,OMIM.ID.supplied.by.OMIM.)
#https://github.com/hbctraining/DGE_workshop/blob/master/lessons/05_DGE_DESeq2_analysis2.md
#convert results from DESeq2 object to data frame and convert rownames to gene column
res<- res %>%
  data.frame() %>%
  rownames_to_column(var="gene_id") %>%
  as_tibble()

view(res)

#this dataframe will be one the hold all the extra info (norm counts, HGNC name, chr, etc)
#now add hgnc_id, gene_name, chr,gene_type
res<-merge(res,gencode_genes,all.x=TRUE)
#now add hgnc stuff
res<-merge(res,hcng_ref,all.x=TRUE,by.x="hgnc_id",by.y="HGNC.ID")
#now order by padj
final_res<-res%>%
  filter(padj<0.05)%>%
  #filter(abs(log2FoldChange)>0.6)%>%
  arrange(desc(abs(log2FoldChange)))
#write.csv(final_res,"sigDEGs FC1.5 Issy ASXL1 blood all samples adjGender.csv")
hist(final_res$log2FoldChange)


summary(res$padj)
```

